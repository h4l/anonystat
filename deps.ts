export { getReasonPhrase, StatusCodes } from "npm:http-status-codes@2";
export { z } from "https://deno.land/x/zod@v3.22.4/mod.ts";
export {
  assert,
  unreachable,
} from "https://deno.land/std@0.221.0/assert/mod.ts";
export {
  difference,
} from "https://deno.land/std@0.221.0/datetime/difference.ts";
export type {
  Unit,
} from "https://deno.land/std@0.221.0/datetime/difference.ts";

export { parse as parseJsonc } from "https://deno.land/std@0.221.0/jsonc/mod.ts";
export type { JsonValue } from "https://deno.land/std@0.221.0/jsonc/mod.ts";
export { generate } from "https://deno.land/std@0.221.0/uuid/v5.ts";
export type { Closer } from "https://deno.land/std@0.221.0/io/types.ts";
