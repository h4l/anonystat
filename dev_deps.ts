export {
  assert,
  assertAlmostEquals,
  assertEquals,
  assertGreater,
  assertInstanceOf,
  AssertionError,
  assertLess,
  assertLessOrEqual,
  assertMatch,
  assertNotEquals,
  assertRejects,
  assertStringIncludes,
} from "https://deno.land/std@0.221.0/assert/mod.ts";

export { assertSnapshot } from "https://deno.land/std@0.221.0/testing/snapshot.ts";

export {
  assertSpyCall,
  spy,
  stub,
} from "https://deno.land/std@0.221.0/testing/mock.ts";

export type { Stub } from "https://deno.land/std@0.221.0/testing/mock.ts";

export { FakeTime } from "https://deno.land/std@0.221.0/testing/time.ts";

import { validate } from "https://deno.land/std@0.221.0/uuid/mod.ts";
export const validateUuid = validate;

export { toText } from "https://deno.land/std@0.221.0/streams/mod.ts";

export { fromFileUrl } from "https://deno.land/std@0.221.0/path/posix/from_file_url.ts";
