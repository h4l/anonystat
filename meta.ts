export enum RequestName {
  collect = "collect",
  debugCollect = "debugCollect",
}

/** Metadata on how an incoming Request matched a handling rule. */
export interface RequestMeta {
  url: URL;
  headers: Headers;
}

export interface DebugRequestMeta {
  debug: boolean;
}

/** The metadata available from GA4 MP request URLs. */
export interface CollectRequestMeta extends RequestMeta, DebugRequestMeta {
  name: RequestName;
  measurement_id: string | null;
  api_secret: string | null;
}

export interface UnknownRequestMeta extends RequestMeta {
  name: null;
}

export type MaybeCollectRequestMeta = CollectRequestMeta | UnknownRequestMeta;

/** The metadata available from GA4 MP request URLs for known, allowed Data Stream. */
export interface ApprovedCollectRequestMeta extends CollectRequestMeta {
  measurement_id: string;
  api_secret: string;
  /** The GA4 MP API URL to send the payload to. */
  endpoint: string;
}
