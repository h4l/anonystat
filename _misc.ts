import { unreachable } from "./deps.ts";

/** Assert that all unions of a type have been [handled in a type-safe way](https://stackoverflow.com/questions/39419170/how-do-i-check-that-a-switch-block-is-exhaustive-in-typescript/39419171#39419171). */
export function assertUnreachable(_value: never, message?: string): never {
  unreachable(message);
}

/** Type-check that an if-else handles all cases, with runtime fallback if not.
 *
 * This allows the type checker to validate that all known types in a union have
 * been handled in a switch/if-else chain, while providing a usable fallback if
 * the type constraint is violated at runtime.
 *
 * @param _value The subject of an if-else chain that is never after the if-else
 * handles all expected values.
 * @param runtimeBehaviour A function that is called at runtime if the static
 * type constraints are violated.
 * @returns The result of runtimeBehaviour.
 */
export function unreachableAtCompileTime<T>(
  _value: never,
  runtimeBehaviour: () => T,
): T {
  return runtimeBehaviour();
}

/** Check if an error is a [TimeoutError](https://developer.mozilla.org/en-US/docs/Web/API/DOMException#timeouterror), e.g. from `AbortSignal.timeout()`.
 */
export function isTimeout(e: unknown): e is DOMException {
  return e instanceof DOMException && e.name == "TimeoutError";
}

/** Check if an error has a message: string property. */
export function hasMessage<T>(e: T): e is T & { message: string } {
  type MaybeMsg = { message?: unknown };
  return e && typeof e === "object" &&
    typeof (e as MaybeMsg).message === "string";
}

export type Error<NameT extends string, DetailsT = unknown> =
  & { name: NameT }
  & DetailsT;

export type SuccessResult<T> = { success: true; data: T };
export type ErrorResult<E> = { success: false; error: E };
export type Result<T, E> = SuccessResult<T> | ErrorResult<E>;

export function mapResult<T, U, E>(
  result: Result<T, E>,
  transform: (value: T) => U,
): Result<U, E> {
  return result.success
    ? { success: true, data: transform(result.data) }
    : result;
}

export type MaybePromise<T> = T | Promise<T>;

// deno-lint-ignore no-explicit-any
export type SyncReturn<T extends (...args: any) => any> = (
  ...args: Parameters<T>
) => Awaited<ReturnType<T>>;

// deno-lint-ignore no-explicit-any
export type IfFn<T, Yes, No> = T extends ((...args: any) => any) ? Yes : No;

export type MaybeLazy<T> = IfFn<T, never, T | (() => T)>;

export function notLazy<T>(maybeLazy: MaybeLazy<T>): T {
  return typeof maybeLazy === "function" ? maybeLazy() : maybeLazy;
}
