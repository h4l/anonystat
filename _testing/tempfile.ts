import { Closer } from "../deps.ts";

/** A reference to a temporary file that's deleted when disposed.
 *
 * ```ts
 * using file = await Deno.open("/foo/bar.txt", { read: true });
 * const fileInfo = await file.stat();
 * if (fileInfo.isFile) {
 *   const buf = new Uint8Array(100);
 *   const numberOfBytesRead = await file.read(buf); // 11 bytes
 *   const text = new TextDecoder().decode(buf);  // "hello world"
 * }
 * ```
 */
export class TemporaryFile implements Closer, Disposable {
  constructor(readonly path: string) {}

  /** Create a TemporaryFile, like {@linkcode Deno.makeTempFile}. */
  static async create(
    options: Deno.MakeTempOptions = {},
  ): Promise<TemporaryFile> {
    return new TemporaryFile(await Deno.makeTempFile(options));
  }

  [Symbol.dispose](): void {
    this.close();
  }

  #closed: boolean = false;
  close(): void {
    if (this.#closed) return;
    this.#closed = true;
    Deno.remove(this.path).catch(() => {});
  }
}
