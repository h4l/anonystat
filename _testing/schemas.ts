import { assert, assertEquals } from "../dev_deps.ts";

import { z } from "../deps.ts";

export function repeat<T>(value: T, times: number): T[] {
  const values = new Array(times);
  values.fill(value);
  return values;
}

export function repeatString(
  value: string,
  times: number,
  joiner: string = "",
): string {
  return repeat(value, times).join(joiner);
}

export const Input = Symbol("Input");

function getInput<Schema extends z.ZodTypeAny>(options: InputOptions<Schema>) {
  return "input" in options ? options.input : options.rawInput;
}

type InputOptions<Schema extends z.ZodTypeAny> = {
  input: z.input<Schema>;
} | { rawInput: unknown };

type MultipleInputOptions<Schema extends z.ZodTypeAny> = {
  inputs: ReadonlyArray<z.input<Schema>>;
} | { rawInputs: ReadonlyArray<unknown> };

function getInputs<Schema extends z.ZodTypeAny>(
  options: MultipleInputOptions<Schema>,
): ReadonlyArray<InputOptions<Schema>> {
  return "inputs" in options
    ? options.inputs.map((i) => ({ input: i }))
    : options.rawInputs.map((ri) => ({ rawInput: ri }));
}

export function assertParses<Schema extends z.ZodTypeAny>(
  { schema, expected = Input, ...inputs }: {
    schema: Schema;
    expected?: z.infer<Schema> | typeof Input;
  } & InputOptions<Schema>,
): void {
  const input = getInput(inputs);
  const result = schema.safeParse(input);
  if (!result.success) {
    assert(
      false,
      `Schema did not parse value: ${
        Deno.inspect({ input, errors: result.error.format()._errors })
      }`,
    );
  }
  assertEquals(result.data, expected === Input ? input : expected);
}

export function assertRejects<Schema extends z.ZodTypeAny>(
  { schema, ...inputs }: {
    schema: Schema;
  } & InputOptions<Schema>,
): void {
  const input = getInput(inputs);
  const result = schema.safeParse(input);
  if (result.success) {
    assert(
      false,
      `Schema parse unexpectedly succeeded: ${
        Deno.inspect({ input, result: result.data })
      }`,
    );
  }
}

export class SchemaAssertions<Schema extends z.ZodTypeAny> {
  constructor(readonly schema: Schema) {}

  assertParses(
    input: z.input<Schema>,
    expected: z.infer<Schema> | typeof Input = Input,
  ): void {
    assertParses({ schema: this.schema, input, expected });
  }

  assertParsesAll(
    options: MultipleInputOptions<Schema>,
  ): void {
    for (const inputOptions of getInputs(options)) {
      assertParses({ schema: this.schema, ...inputOptions });
    }
  }

  assertRejects(
    input: z.input<Schema>,
  ): void {
    assertRejects({ schema: this.schema, input });
  }

  assertRejectsAll(
    options: MultipleInputOptions<Schema>,
  ): void {
    for (const inputOptions of getInputs(options)) {
      assertRejects({ schema: this.schema, ...inputOptions });
    }
  }
}
