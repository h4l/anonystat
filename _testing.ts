import { string } from "https://deno.land/x/zod@v3.22.4/types.ts";
import { ErrorResult, Result, SuccessResult } from "./_misc.ts";
import {
  assert,
  assertEquals,
  AssertionError,
  assertLess,
  assertLessOrEqual,
  Stub,
  validateUuid,
} from "./dev_deps.ts";
import { assertFalse } from "https://deno.land/std@0.221.0/assert/assert_false.ts";
import { AnyPayload } from "./payload-schemas.ts";
import { assertGreater } from "./dev_deps.ts";

/** Parse an iso date and return it as a ms timestamp, throw on invalid syntax. */
export function timestamp(isoDate: string): number {
  const timestamp = Date.parse(isoDate);
  if (Number.isNaN(timestamp)) throw new Error("Invalid date");
  return timestamp;
}

/** Parse an iso date and return it as a Date, throw on invalid syntax. */
export function date(isoDate: string, offset: number = 0): Date {
  return new Date(timestamp(isoDate) + offset);
}

export function assertSuccessful<T, E>(
  result: Result<T, E>,
): asserts result is SuccessResult<T> {
  if (result.success) return;
  assert(
    false,
    `expected result to be successful but it failed with error: ${
      Deno.inspect(result.error)
    }`,
  );
}

export function assertUnsuccessful<T, E>(
  result: Result<T, E>,
): asserts result is ErrorResult<E> {
  if (!result.success) return;
  assert(
    false,
    `result is not unsuccessful with data: ${Deno.inspect(result.data)}`,
  );
}

/** Assert that value is a valid UUID string. */
export function assertUuid(value: unknown): asserts value is string {
  if (typeof value === "string" && validateUuid(value)) return;
  assert(false, `value is not a UUID: ${Deno.inspect(value)}`);
}

/** Linearly interpolate from `a` to `b` by `t`. */
export function lerp(a: number, b: number, t: number): number {
  return a + (b - a) * t;
}

export function assertResponseOk(response: Response): void {
  if (response.status >= 200 && response.status < 300) return;
  assert(
    false,
    `expected response to have 2XX status but is ${response.status}: ${
      Deno.inspect(response)
    }`,
  );
}

export async function assertOneFetchForMpCollectRequest(
  fetch: Stub<
    typeof globalThis,
    [input: string | Request | URL, init?: RequestInit | undefined],
    Promise<Response>
  >,
  mpCollectRequestAttrs: MpCollectRequestAttrs,
) {
  await assertFetchForMpCollectRequest(
    fetch,
    { index: 0, maxCalls: 1 },
    mpCollectRequestAttrs,
  );
}
export async function assertFetchForMpCollectRequest(
  fetch: Stub<
    typeof globalThis,
    [input: string | Request | URL, init?: RequestInit | undefined],
    Promise<Response>
  >,
  { index, maxCalls }: { index: number; maxCalls?: number },
  mpCollectRequestAttrs: MpCollectRequestAttrs,
) {
  if (maxCalls !== undefined) {
    assertLessOrEqual(fetch.calls.length, maxCalls, "Too many fetch() calls");
  }
  assertGreater(fetch.calls.length, index, "Too few fetch() calls");

  const request = fetch.calls[index].args[0];
  if (
    !(fetch.calls[index].args.length === 1 && request instanceof Request)
  ) {
    assert(
      false,
      `expected fetch to be called with one Request argument: ${
        Deno.inspect(fetch.calls[index].args)
      }`,
    );
  }
  await assertIsMpRequest(request, mpCollectRequestAttrs);
}

export type MpCollectRequestAttrs = {
  origin: string;
  measurement_id: string;
  api_secret: string;
  user_id?: string | ((value: string | undefined) => void);
};

export async function assertIsMpRequest(
  request: Request,
  { origin, measurement_id, api_secret, user_id }: MpCollectRequestAttrs,
): Promise<void> {
  const reqUrl = new URL(request.url);
  assertEquals(reqUrl.origin, origin);
  assertEquals(reqUrl.pathname, "/mp/collect");
  assertEquals(reqUrl.searchParams.get("measurement_id"), measurement_id);
  assertEquals(reqUrl.searchParams.get("api_secret"), api_secret);

  if (user_id !== undefined) {
    const payload = await getRequestPayload(request);
    if (typeof user_id === "string") {
      assertEquals(
        payload.user_id,
        user_id,
        "Payload contains unexpected user_id",
      );
    } else {
      user_id(payload.user_id);
    }
  }
}

async function getRequestPayload(request: Request): Promise<AnyPayload> {
  try {
    return AnyPayload.parse(await request.json());
  } catch (e) {
    throw new AssertionError(
      `Request body is not readable as a collect request payload: ${
        Deno.inspect(e)
      }`,
    );
  }
}
