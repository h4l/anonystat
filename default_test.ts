import { ErrorResult } from "./_misc.ts";
import { assertSuccessful, assertUnsuccessful } from "./_testing.ts";
import {
  approvedRequestDestinationSelector,
  createPayloadParser,
  createProxySender,
  defaultPayloadParser,
  defaultProxySender,
  defaultRequestReader,
  defaultResponseWriter,
  GA4MPDestination,
  matchDefaultGA4MPUrls,
} from "./default.ts";
import { assert, StatusCodes, z } from "./deps.ts";
import { assertEquals, assertSnapshot, stub, toText } from "./dev_deps.ts";
import {
  ApprovedCollectRequestMeta,
  DebugRequestMeta,
  RequestMeta,
} from "./meta.ts";
import { RequestName } from "./meta.ts";
import { AnyPayload } from "./payload-schemas.ts";
import {
  GA4MPPayload,
  MatchedErrorNames,
  PayloadParseError,
  PayloadParser,
  ProxySender,
  ProxySendError,
  RequestReadError,
  UnknownPayload,
} from "./proxy.ts";

Deno.test("matchDefaultGA4MPUrls()", async (t) => {
  const completed = Promise.reject("not used");
  completed.catch(() => {}); // ignore

  const info: Deno.ServeHandlerInfo = {
    completed,
    remoteAddr: { hostname: "example", port: 12345, transport: "tcp" },
  };

  function req(url: string): [request: Request, info: Deno.ServeHandlerInfo] {
    return [new Request(url, { headers: { "x-foo": "bar" } }), info];
  }

  await t.step("Does not match unknown request", () => {
    const url = "https://example.com/";
    const [request, info] = req(url);
    const result = matchDefaultGA4MPUrls(request, info);

    assertEquals(result.name, null);
    assertEquals(result.url, new URL(url));
    assertEquals(result.headers, request.headers);
  });

  await t.step("Matches collect request", () => {
    const url =
      "https://example.com/mp/collect?measurement_id=foo&api_secret=bar";
    const [request, info] = req(url);
    const result = matchDefaultGA4MPUrls(request, info);

    assertEquals(result.url, new URL(url));
    assertEquals(result.headers, request.headers);
    assert(result.name === RequestName.collect);
    assertEquals(result.debug, false);
    assertEquals(result.measurement_id, "foo");
    assertEquals(result.api_secret, "bar");
  });

  await t.step("Matches debug collect request", () => {
    const url =
      "https://example.com/debug/mp/collect?measurement_id=foo&api_secret=bar";
    const [request, info] = req(url);
    const result = matchDefaultGA4MPUrls(request, info);

    assertEquals(result.url, new URL(url));
    assertEquals(result.headers, request.headers);
    assert(result.name === RequestName.debugCollect);
    assertEquals(result.debug, true);
    assertEquals(result.measurement_id, "foo");
    assertEquals(result.api_secret, "bar");
  });
});

Deno.test("defaultRequestReader()", async (t) => {
  const url =
    "https://example.com/mp/collect?measurement_id=foo&api_secret=bar";

  const payload: AnyPayload = { client_id: "test", events: [] };

  function createRequest(
    {
      method = "POST",
      contentType = "application/json",
      body = JSON.stringify(payload),
      ioError = false,
    }: {
      method?: string;
      contentType?: string;
      body?: string;
      ioError?: boolean;
    } = {},
  ): [request: Request, meta: RequestMeta] {
    const headers = new Headers({
      "content-type": contentType,
    });
    const request = new Request(url, { method, headers, body });
    if (ioError) {
      stub(request, "json", () => {
        throw new Deno.errors.ConnectionReset();
      });
    }
    return [request, { url: new URL(url), headers }];
  }

  await t.step("reads valid request", async () => {
    const [request, requestMeta] = createRequest();
    const result = await defaultRequestReader(request, { requestMeta });
    assertSuccessful(result);
    assertEquals(result.data.payload, payload);
  });

  await t.step("rejects requests with methods other than POST", async () => {
    const [request, requestMeta] = createRequest({ method: "PUT" });
    const result = await defaultRequestReader(request, { requestMeta });
    assertUnsuccessful(result);
    assertEquals(result.error.name, "incorrect-request-method");
  });

  await t.step("rejects requests without JSON content type", async () => {
    const [request, requestMeta] = createRequest({ contentType: "foo/bar" });
    const result = await defaultRequestReader(request, { requestMeta });
    assertUnsuccessful(result);
    assertEquals(result.error.name, "incorrect-content-type");
  });

  await t.step("rejects requests with syntactically-invalid JSON", async () => {
    const [request, requestMeta] = createRequest({ body: "{" });
    const result = await defaultRequestReader(request, { requestMeta });
    assertUnsuccessful(result);
    assertEquals(result.error.name, "body-not-valid-json");
  });

  await t.step("rejects requests that cannot be read", async () => {
    const [request, requestMeta] = createRequest({ ioError: true });
    const result = await defaultRequestReader(request, { requestMeta });
    assertUnsuccessful(result);
    assertEquals(result.error.name, "request-io-error");
  });
});

Deno.test("defaultResponseWriter()", async (t) => {
  const requestMeta: RequestMeta = {
    url: new URL("https://example.com/mp/collect"),
    headers: new Headers(),
  };
  const debugRequestMeta: RequestMeta & DebugRequestMeta = {
    ...requestMeta,
    debug: true,
  };

  await t.step("success sends no content", async () => {
    const response = await defaultResponseWriter({
      success: true,
      data: { payload: { payload: {} }, proxyResult: {} },
    }, { requestMeta });

    assertEquals(response.status, StatusCodes.NO_CONTENT);
    assertEquals(response.body, null);
  });

  type ForwardingError = RequestReadError | PayloadParseError | ProxySendError;

  const errorStatuses: [error: ForwardingError, status: number][] = [
    [{ name: "aborted" }, StatusCodes.INTERNAL_SERVER_ERROR],
    [{ name: "body-not-valid-json" }, StatusCodes.BAD_REQUEST],
    [{ name: "incorrect-content-type" }, StatusCodes.NOT_ACCEPTABLE],
    [{ name: "incorrect-request-method" }, StatusCodes.METHOD_NOT_ALLOWED],
    [
      { name: "invalid-ga4mp-payload", zodError: new z.ZodError([]) },
      StatusCodes.BAD_REQUEST,
    ],
    [{ name: "proxy-io-error" }, StatusCodes.BAD_GATEWAY],
    [{ name: "proxy-response-status", status: 500 }, StatusCodes.BAD_GATEWAY],
    [{ name: "request-io-error" }, StatusCodes.BAD_REQUEST],
    [{ name: "timeout" }, StatusCodes.GATEWAY_TIMEOUT],
  ];

  assertEquals(
    new Set(errorStatuses.map(([e]) => e.name)),
    new Set(Object.values(MatchedErrorNames.Enum)),
    "errors test cases don't match known error names",
  );

  for (const [error, status] of errorStatuses) {
    await t.step(
      `error ${error.name} response with status ${status}`,
      async () => {
        const response = await defaultResponseWriter(
          { success: false, error },
          { requestMeta },
        );

        assertEquals(response.status, status);
      },
    );
  }

  await t.step("invalid-ga4mp-payload response", async (t) => {
    function invalidPayloadError(): ErrorResult<PayloadParseError> {
      return {
        success: false,
        error: {
          name: "invalid-ga4mp-payload",
          zodError: new z.ZodError([{
            code: "invalid_type",
            expected: "string",
            received: "array",
            path: ["user_id"],
            message: "Must be a string",
          }]),
        },
      };
    }

    await t.step("contains details of error in debug mode", async (t) => {
      const response = await defaultResponseWriter(invalidPayloadError(), {
        requestMeta: debugRequestMeta,
      });

      assertEquals(response.status, StatusCodes.BAD_REQUEST);
      assert(response.body);
      await assertSnapshot(t, await toText(response.body));
    });

    await t.step("omits details of error in non-debug mode", async () => {
      const response = await defaultResponseWriter(invalidPayloadError(), {
        requestMeta, // non-debug
      });

      assertEquals(response.status, StatusCodes.BAD_REQUEST);
      assert(response.body);
      assertEquals(
        await toText(response.body),
        "Request body is not a valid GA4 Measurement Protocol payload",
      );
    });
  });

  await t.step("handles unexpected error values", async () => {
    const response = await defaultResponseWriter(
      // deno-lint-ignore no-explicit-any
      { success: false, error: {} as any },
      { requestMeta },
    );

    assertEquals(response.status, StatusCodes.INTERNAL_SERVER_ERROR);
    assert(response.body);
    assertEquals(
      await toText(response.body),
      "Unknown error: undefined",
    );
  });
});

Deno.test("approvedRequestDestinationSelector()", async () => {
  const payload: UnknownPayload = { payload: {} };
  const endpoint = "https://dest.example.com/mp/collect";
  const measurement_id = "exampleId";
  const api_secret = "exampleSec";
  const requestMeta: ApprovedCollectRequestMeta = {
    url: new URL("https://example.com/mp/collect"),
    headers: new Headers(),
    name: RequestName.collect,
    debug: false,
    measurement_id,
    api_secret,
    endpoint,
  };

  const destination: GA4MPDestination = approvedRequestDestinationSelector({
    payload,
    requestMeta,
  });

  assertEquals(destination, { measurement_id, api_secret, endpoint });
});

async function testPayloadParser<T>(
  t: Deno.TestContext,
  { validInput, validOutput, invalidInput, parseError, payloadParser }: {
    validInput: unknown;
    validOutput: T;
    invalidInput: unknown;
    parseError: string;
    payloadParser: PayloadParser<
      UnknownPayload,
      GA4MPPayload<T>,
      PayloadParseError,
      RequestMeta
    >;
  },
) {
  const requestMeta: RequestMeta = {
    url: new URL("https://example.com/mp/collect"),
    headers: new Headers(),
  };

  await t.step("accepts payload matching schema", async () => {
    const result = await payloadParser({ payload: validInput }, {
      requestMeta,
    });
    assertSuccessful(result);
    assertEquals(result.data, { payload: validOutput });
  });

  await t.step("rejects payload not matching schema", async () => {
    const result = await payloadParser({ payload: invalidInput }, {
      requestMeta,
    });
    assertUnsuccessful(result);
    assertEquals(result.error.name, "invalid-ga4mp-payload");

    assertEquals(result.error.zodError.issues.map((i) => i.message), [
      parseError,
    ]);
  });
}

Deno.test("createPayloadParser()", async (t) => {
  const payloadParser = createPayloadParser(
    z.object({ foo: z.string().pipe(z.coerce.number()) }),
  );

  await testPayloadParser(t, {
    payloadParser,
    validInput: { foo: "42" },
    validOutput: { foo: 42 },
    invalidInput: { foo: true },
    parseError: "Expected string, received boolean",
  });
});

Deno.test("defaultPayloadParser()", async (t) => {
  await testPayloadParser(t, {
    payloadParser: defaultPayloadParser,
    validInput: { client_id: "test", events: [], ignored: true },
    validOutput: { client_id: "test", events: [] },
    invalidInput: { client_id: "test", events: {} },
    parseError: "Expected array, received object",
  });
});

function nextAbortEvent(signal: AbortSignal): Promise<Event> {
  return new Promise((resolve) => {
    signal.addEventListener("abort", resolve);
  });
}

function stubFetchToBlockUntilAbort() {
  return stub(
    globalThis,
    "fetch",
    async (input, init = {}) => {
      const signal = input instanceof Request ? input.signal : init.signal;
      assert(signal, "fetch() called without a signal");
      await nextAbortEvent(signal);
      signal.throwIfAborted();
      assert(false, "signal did not throw");
    },
  );
}

async function testProxySender<
  PayloadT extends UnknownPayload,
  ProxySendResultT,
  ProxySendErrorT extends ProxySendError,
  RequestMetaT extends RequestMeta,
>(t: Deno.TestContext, { proxySender, validPayload, requestMeta, sendResult }: {
  proxySender: ProxySender<
    PayloadT,
    ProxySendResultT,
    ProxySendError | ProxySendErrorT,
    RequestMetaT
  >;
  validPayload: PayloadT;
  requestMeta: RequestMetaT;
  sendResult: ProxySendResultT;
}) {
  await t.step(
    "returns result from resultCreator when fetch succeeds",
    async () => {
      using fetchStub = stub(globalThis, "fetch", async () => {
        return new Response(null, { status: StatusCodes.NO_CONTENT });
      });
      const result = await proxySender(validPayload, { requestMeta });
      assertSuccessful(result);
      assertEquals(result.data, sendResult);
      assertEquals(fetchStub.calls.length, 1);
    },
  );

  await t.step(
    "cancels fetch and returns timeout when signal times out",
    async () => {
      const timeout = AbortSignal.timeout(0);
      using fetchStub = stubFetchToBlockUntilAbort();

      const resultP = proxySender(validPayload, {
        requestMeta,
        signal: timeout,
      });
      const result = await resultP;

      assertEquals(fetchStub.calls.length, 1);
      assertUnsuccessful(result);
      assertEquals(result.error.name, "timeout");
    },
  );

  await t.step(
    "cancels fetch and returns aborted error when signal aborts",
    async () => {
      const abort = new AbortController();
      using fetchStub = stubFetchToBlockUntilAbort();

      const resultP = proxySender(validPayload, {
        requestMeta,
        signal: abort.signal,
      });
      abort.abort();
      const result = await resultP;

      assertEquals(fetchStub.calls.length, 1);
      assertUnsuccessful(result);
      assertEquals(result.error.name, "aborted");
    },
  );

  await t.step(
    "returns proxy-io-error error when fetch fails",
    async () => {
      using fetchStub = stub(globalThis, "fetch", async () => {
        // fetch() is specc'd to throw TypeError for network errors:
        // See 5.6 / 12 / 3 in https://fetch.spec.whatwg.org/#fetch-method
        throw new TypeError("network error");
      });

      const result = await proxySender(validPayload, { requestMeta });

      assertEquals(fetchStub.calls.length, 1);
      assertUnsuccessful(result);
      assert(result.error.name === "proxy-io-error");
      assertEquals(result.error.message, "network error");
    },
  );

  await t.step(
    "returns proxy-response-status error when fetch response is not ok",
    async () => {
      using fetchStub = stub(globalThis, "fetch", async () => {
        return new Response(null, {
          status: StatusCodes.IM_A_TEAPOT,
        });
      });

      const result = await proxySender(validPayload, { requestMeta });

      assertEquals(fetchStub.calls.length, 1);
      assertUnsuccessful(result);
      assert(result.error.name === "proxy-response-status");
      assertEquals(result.error.status, StatusCodes.IM_A_TEAPOT);
    },
  );
}

Deno.test("createProxySender()", async (t) => {
  const payload: UnknownPayload = { payload: {} };
  const endpoint = "https://dest.example.com/mp/collect";
  const measurement_id = "exampleId";
  const api_secret = "exampleSec";
  const requestMeta: ApprovedCollectRequestMeta = {
    url: new URL("https://example.com/mp/collect"),
    headers: new Headers(),
    name: RequestName.collect,
    debug: false,
    measurement_id,
    api_secret,
    endpoint,
  };

  const proxySender = createProxySender({
    destinationSelector(options): GA4MPDestination {
      assertEquals(options.payload, payload);
      assertEquals(options.requestMeta, requestMeta);

      return {
        measurement_id: "id",
        api_secret: "sec",
        endpoint: "https://example.com/mp/collect",
      };
    },
    resultCreator({ request, response, error, ...options }) {
      assertEquals(options.payload, payload);
      assertEquals(options.requestMeta, requestMeta);

      assertEquals(
        request.url,
        "https://example.com/mp/collect?api_secret=sec&measurement_id=id",
      );

      if (error === null) {
        assertEquals(response?.status, StatusCodes.NO_CONTENT);
        return { success: true, data: "foo" };
      }

      return { success: false, error };
    },
  });

  await testProxySender(t, {
    proxySender,
    requestMeta,
    sendResult: "foo",
    validPayload: payload,
  });
});

Deno.test("defaultProxySender()", async (t) => {
  const requestMeta: ApprovedCollectRequestMeta = {
    url: new URL("https://example.com/mp/collect"),
    headers: new Headers(),
    name: RequestName.collect,
    debug: false,
    measurement_id: "exampleId",
    api_secret: "exampleSec",
    endpoint: "https://dest.example.com/mp/collect",
  };

  await testProxySender(t, {
    proxySender: defaultProxySender,
    requestMeta,
    sendResult: null,
    validPayload: {
      payload: { client_id: "test", events: [] } satisfies AnyPayload,
    },
  });
});
