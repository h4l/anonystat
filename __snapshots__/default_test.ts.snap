export const snapshot = {};

snapshot[`defaultResponseWriter() > invalid-ga4mp-payload response > contains details of error in debug mode 1`] = `
'Request body is not a valid GA4 Measurement Protocol payload: [
  {
    code: "invalid_type",
    expected: "string",
    received: "array",
    path: [ "user_id" ],
    message: "Must be a string"
  }
]'
`;
